homeassistant/components/http/__init__.py

```diff
@@ -460,21 +460,26 @@ class RequestDataValidator:
     Will return a 400 if no JSON provided or doesn't match schema.
     """
 
-    def __init__(self, schema):
+    def __init__(self, schema, allow_empty=False):
         """Initialize the decorator."""
         self._schema = schema
+        self._allow_empty = allow_empty
 
     def __call__(self, method):
         """Decorate a function."""
         @asyncio.coroutine
         @wraps(method)
         def wrapper(view, request, *args, **kwargs):
             """Wrap a request handler with data validation."""
+            data = None
             try:
                 data = yield from request.json()
             except ValueError:
-                _LOGGER.error('Invalid JSON received.')
-                return view.json_message('Invalid JSON.', 400)
+                if not self._allow_empty or \
+                   (yield from request.content.read()) != b'':
```

Comment:
You can realy read 2 times the same body?

Reply:
I've opened a separate PR with this change and added tests. Will rebase htis PR after it is merged https://github.com/home-assistant/home-assistant/pull/12401